numpyro/util.py

```diff
@@ -121,21 +121,25 @@ def _identity(x):
 
 def tscan(f, a, bs, transform=_identity):
     if _DISABLE_CONTROL_FLOW_PRIM:
-        length = tree_flatten(bs)[0][0].shape[0]
-        for i in range(length):
-            b = tree_map(lambda x: x[i], bs)
-            a = f(a, b)
-            a_out = transform(tree_map(lambda x: np.expand_dims(x, axis=0), a))
-            if i == 0:
-                out = a_out
-            else:
-                out = tree_multimap(lambda x, y: np.concatenate((x, y)) if x is not None else None,
-                                    out, a_out)
-        return out
+        tscan_nonprim(f, a, bs, transform=_identity)
```

Comment:
nit: append underscore to deter usage outside this file.

Reply:
Do you mean `_tscan_nonprim`? Indeed, I use this implementation directly (not through tscan because adding a nonprim context manager will force other primitives become nonprim too). ^^!

Reply:
Hmm..I thought you were running some other custom implementation. :) So you are saying that this naive version is faster than the actual `lax.scan` version? Why is that the case - is it just due to the initial small step size?

Reply:
> is it just due to the initial small step size?

yup, I think so. `jit(sample_kernel)(hmc_sate)` took more than 1 hour in my system while `jit(sample_kernel)(hmc_state.update(step_size=1.))` took me 15s. >"<

Reply:
I'm fine with merging this then, but would like to find a way to resolve this within `tscan` itself finally.

Reply:
I think we can use `tscan` for large models again when issues with first step of while_loop is resolved. :)

Hmm, I just thought that my suggestions are not good. I should make a custom tscan in my notebook instead, where I can add print statements too. Sorry for wandering around!